src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs

```diff
@@ -277,28 +199,131 @@ public override void WriteTo(TextWriter writer, HtmlEncoder encoder)
+                encoder.Encode(writer, stringValue);
+            }
+            else
+            {
+                ((IHtmlContent)entry).WriteTo(writer, encoder);
+            }
+        }
 
-                var stringValue = entry as string;
-                if (stringValue != null)
+        private void CopyToCore(object entry, IHtmlContentBuilder destination)
+        {
+            if (entry == null)
+            {
+                return;
+            }
+
+            string entryAsString;
+            IHtmlContentContainer entryAsContainer;
+            if ((entryAsString = entry as string) != null)
+            {
+                destination.Append(entryAsString);
+            }
+            else if ((entryAsContainer = entry as IHtmlContentContainer) != null)
+            {
+                entryAsContainer.CopyTo(destination);
+            }
+            else
+            {
+                destination.AppendHtml((IHtmlContent)entry);
+            }
+        }
+
+        private void MoveToCore(object entry, IHtmlContentBuilder destination)
+        {
+            if (entry == null)
+            {
+                return;
+            }
+
+            string entryAsString;
+            IHtmlContentContainer entryAsContainer;
+            if ((entryAsString = entry as string) != null)
+            {
+                destination.Append(entryAsString);
+            }
+            else if ((entryAsContainer = entry as IHtmlContentContainer) != null)
+            {
+                entryAsContainer.MoveTo(destination);
+            }
+            else
+            {
+                destination.AppendHtml((IHtmlContent)entry);
+            }
+        }
+
+        private bool IsEmptyOrWhiteSpaceCore(object entry, EmptyOrWhiteSpaceWriter writer)
+        {
+            if (entry == null)
+            {
+                return false;
+            }
+
+            var stringValue = entry as string;
+            if (stringValue != null)
+            {
+                if (!string.IsNullOrWhiteSpace(stringValue))
                 {
-                    encoder.Encode(writer, stringValue);
+                    return false;
                 }
-                else
+            }
+            else
+            {
+                ((IHtmlContent)entry).WriteTo(writer, HtmlEncoder.Default);
+                if (!writer.IsEmptyOrWhiteSpace)
                 {
-                    ((IHtmlContent)entry).WriteTo(writer, encoder);
+                    return false;
                 }
             }
+
+            return true;
+        }
+
+        private TagHelperContent AppendCore(object entry)
+        {
+            if (!_hasContent)
+            {
+                _hasContent = true;
+                _isModified = true;
+                _isSingleContentSet = true;
+                _singleContent = entry;
+            }
+            else
+            {
+                Buffer.Add(entry);
+
+                _hasContent = _buffer.Count > 0;
```

Comment:
Huh? `Buffer.Count > 1` is an invariant at this line. Just set `_hasContent` and `isModified` to `true` prior to the `if`.